src/ucp/core/ucp_device.c

```diff
@@ -314,8 +324,38 @@ static ucs_status_t ucp_device_mem_list_create_handle(
         return status;
     }
 
+    /* populate elements common parameters */
+    local_addresses = (void**)UCS_PTR_BYTE_OFFSET(mem->address, sizeof(handle));
+    remote_addresses = (uint64_t*)UCS_PTR_BYTE_OFFSET(
+            local_addresses,
+            sizeof(*handle.ucp_mem_elements.local_addr) * params->num_elements);
+    lengths          = (size_t*)UCS_PTR_BYTE_OFFSET(
+            remote_addresses, sizeof(*handle.ucp_mem_elements.remote_addr) *
+                                      params->num_elements);
+    for (i = 0; i < params->num_elements; i++) {
+        ucp_element = &params->elements[i];
+        local_addr  = UCS_PARAM_VALUE(UCP_DEVICE_MEM_LIST_ELEM_FIELD,
+                                      ucp_element, local_addr, LOCAL_ADDR, NULL);
+        remote_addr = UCS_PARAM_VALUE(UCP_DEVICE_MEM_LIST_ELEM_FIELD,
+                                      ucp_element, remote_addr, REMOTE_ADDR, 0);
+        length = UCS_PARAM_VALUE(UCP_DEVICE_MEM_LIST_ELEM_FIELD, ucp_element,
+                                 length, LENGTH, 0);
+        ucp_mem_type_unpack(ep->worker, &local_addresses[i], &local_addr,
+                            sizeof(local_addresses[i]), mem_type);
+        ucp_mem_type_unpack(ep->worker, &remote_addresses[i], &remote_addr,
+                            sizeof(remote_addresses[i]), mem_type);
+        ucp_mem_type_unpack(ep->worker, &lengths[i], &length,
+                            sizeof(lengths[i]), mem_type);
```

Comment:
can we "unpack" all these in single/several calls, and not for each element?

Reply:
remote_addrs array follow local_addrs array, so only way I see to do it not for each element is to allocate some tmp array, copy to it and then use unpack.